astarte/device/device.py

```diff
@@ -369,6 +372,14 @@ def __on_disconnect(self, client, userdata, rc):
         # If rc was explicit, stop the loop (after the callback)
         if rc == 0:
             self.__mqtt_client.loop_stop()
+        # Else check certificate expiration and try reconnection
+        elif crypto.certificate_is_expired(os.path.join(self.__persistency_dir, self.__device_id,
+                                                        "crypto")):
```

Comment:
The `certificate_is_expired` check is legit when it's performed proactively (e.g. at the device start), but here the correct check is probably `rc == MQTT_ERR_TLS` (see [here](https://github.com/eclipse/paho.mqtt.python/blob/9c22a9c297c0cdc4e1aac13aa19073e09a822961/src/paho/mqtt/client.py#L157)).

This is because the certificate can be invalid for various reasons: the CA can be changed, the client and the server can have different local times, etc., so usually SDKs react to TLS connection errors requesting a new certificate.

Reply:
After checking the code and the behaviour of the software I can assure you that the rc is always 1 in this scenario

Reply:
If the library doesn't surface the TLS error correctly then the above heuristic is better than nothing, but please add a `TODO: check for MQTT_ERR_TLS when Paho correctly returns it` so we can return on this later if the library gets fixed.

It would also be a good idea to provide a minimal example reproducing this to open an issue on the Paho repository.

Reply:
Totally reasonable. TODO Added.